lib/web/resources.go

```diff
@@ -78,22 +78,47 @@ func (h *Handler) checkAccessToRegisteredResource(w http.ResponseWriter, r *http
 	}, nil
 }
 
-func (h *Handler) getRolesHandle(w http.ResponseWriter, r *http.Request, params httprouter.Params, ctx *SessionContext) (interface{}, error) {
+func (h *Handler) listRolesHandle(w http.ResponseWriter, r *http.Request, params httprouter.Params, ctx *SessionContext) (interface{}, error) {
 	clt, err := ctx.GetClient()
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	return getRoles(clt)
+	values := r.URL.Query()
+	return listRoles(clt, values)
 }
 
-func getRoles(clt resourcesAPIGetter) ([]ui.ResourceItem, error) {
-	roles, err := clt.GetRoles(context.TODO())
+func listRoles(clt resourcesAPIGetter, values url.Values) (*listResourcesWithoutCountGetResponse, error) {
+	limit, err := QueryLimitAsInt32(values, "limit", defaults.MaxIterationLimit)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	roles, err := clt.ListRoles(context.TODO(), &proto.ListRolesRequest{
+		Limit:    limit,
+		StartKey: values.Get("startKey"),
+		Filter: &types.RoleFilter{
+			SearchKeywords: client.ParseSearchKeywords(values.Get("search"), ' '),
+		},
+	})
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	return ui.NewRoles(roles)
+	var typeRoles []types.Role
+	for _, role := range roles.GetRoles() {
+		typeRoles = append(typeRoles, role)
+	}
+
+	uiRoles, err := ui.NewRoles(typeRoles)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &listResourcesWithoutCountGetResponse{
+		Items:    uiRoles,
+		StartKey: roles.GetNextKey(),
+	}, nil
```

Comment:
We need to be able to handle requests from an old webclient (in the case of two different proxies running together) which would still expect a response that is only an array of items, rather than this paginated object.

Without having to resort to a larger refactor that can version this endpoint, we can just use the existence of one of these new parameters to determine between old/new clients. For example "if `limit` exists, send new response. otherwise, send only `uiRoles`"

We did something similar [here](https://github.com/gravitational/teleport.e/pull/3738/files#diff-33313ae55ac65365e4e8da67e6ad1107065f983e6e20d4dd8119ace20d1284d8R393-R401)

Reply:
Also, make sure the `TODO` i left in `useKeyBasedPagination` would fix this from the frontend side as well please (new webclient being able to handle "old" response types).

Reply:
Oh, I completely forgot that we can have proxies running different versions. 
I added a check on the parameters to determine what response type should be returned. It's not beautiful but works.

>Also, make sure the TODO i left in useKeyBasedPagination would fix this from the frontend side as well please 

The thing I could do to allow new webclient being able to handle "old" response types, would be to wrap the response in the service layer if I detect that I didn't receive the paginated object. But I'm not sure if it makes any sense, things like `limit` or `search` would not work at all. 

I don't know how to handle this, to be honest. If I have a server-side table, how can I make it work with both paginated and non-paginated responses? 
Maybe instead of modifying the existing endpoint I should add a new one? At least the request would fail with 404 error instead of something like `res.map is not a function`.

Reply:
> I don't know how to handle this, to be honest. If I have a server-side table, how can I make it work with both paginated and non-paginated responses?

The worst case scenario is we receive a very large dataset from the "old" api. I don't have any answer for this yet. An upcoming project will be to add a properly versioned API. I don't know exactly how it would work yet for newer web clients receiving 404s but my guess is there would have to be a fallback to a previous version or just accept the 404. I want to discuss this with the team with an RFD once we get the greenlight for the project.

Reply:
Actually, we can support the "new webclient being able to handle old response types" case too.
I added a compatibility layer in `resource.ts`: if we detect that we received a non-paginated response, we can "paginate" it locally, in the browser. I added support for `search`, `limit` and `startKey`.
Let me know what you think about this.

Reply:
i LOVE this. amazing work!!